File: scala_version/spark_SQL_DataFrames_Datasets.scala
// SQL 
	// Spark SQL => execute SQL queries
	// Spark SQL => read data from an existing Hive installation
	// Spark SQL => run SQL from within another programming language(scala, java, python) 
		// Dataset/DataFrame
	// Spark SQL => interact with SQL interface
		// Spark SQL CLI: spark-sql
		// JDBC/ODBC

// Datasets and DataFrame
// Dataset 
	// distributed collection of data
	// optimized RDD
	// scala
// DataFrame
	// Dataset organized into named columns => table/dataframe
	// sources:
		// structured data files
		// Hive table
		// external databases
		// existing RDD
	// scala, python, R
		// scala: Dataset of RowS: Dataset[Row]
		// Python: DataFrame
		// R: data.frame





// ===============================================================
// Spark SQL
// ===============================================================
import org.apache.spark.sql.SparkSession
import spark.implicits._

object spark_SQL_DataFrames_Datasets {
	def main(args: Array[String]) {

		// ------------------------------------------------------
		// 程序入口: SparkSession
		// ------------------------------------------------------
		val spark = SparkSession
			.builder()
			.appName("Spark SQL")
			.config("spark.some.config.option", "some-value")
			.getOrCreate()

		// -------------------------------------------------------
		// 创建 DataFrames
		// existing RDD, Hive table, Spark data sources
		// -------------------------------------------------------
		val df = spark.read.json("E:/DataScience/spark_data/data/examples/src/main/resources/people.json")
		df.show()

		// ------------------------------------------------------
		// Untyped(弱类型) Dataset Operations(aka DataFrame Operations)
		// DataFrames are just Dataset of Rows in Scala and Java API. 
		// These operations are also referred as “untyped transformations” 
		// in contrast to “typed transformations” come with strongly typed Scala/Java Datasets.
		// ------------------------------------------------------
		df.printSchema() 						// print the shcema in a tree format
		df.select("name").show()				// select only the "name" column
		df.select($"name", $"age" + 1).show()	// select name and age, increment the age by 1
		df.filter($"age" > 21).show()			// select people older than 21
		df.groupBy("age").count().show()		// Count people by age




		// --------------------------------
		// Running SQL Queries Programmatically
		// --------------------------------
		// Temporary View
		df.createOrReplaceTempView("people")
		val sqlDF = spark.sql("SELECT * FROM people")
		sqlDF.show()

		// Global temporary view
		df.createGlobalTempView("people")
		val sqlDF_Global = spark.sql("SELECT * FROM global_temp.people")
		sqlDF_Global.show()

		// Global temporary view is cross-session
		spark.newSession().sql("SELECT * FROM global_temp.people").show()

		// --------------------------------
		// Create Datasets
		// --------------------------------
		case class Person(name: String, age: Long)

		// Encoders are created for case classes
		val caseClassDS = Seq(Person("Andy", 32)).toDS()
		caseClassDS.show()

		// Encoders for most common types are automatically provided by importing spark.implicits._
		val primitiveDS = Seq(1, 2, 3).toDS()
		primitiveDS.map(_ + 1).collect()

		// DataFrame can be converted to a Dataset by providing a class.
		val path = "D:/spark/example/src/main/resources/people.json"
		val peopleDS = spark.read.json(path).as[Person]
		peopleDS.show()

		// --------------------------------
		// Dataset and RDD
		// Inferring the Schema Using Reflection

		// Programmatically Specifying the Schema

		// --------------------------------





		// stop the spark
		spark.stop()
	}
}







// 聚合操作
// Untyped User-Defined Aggregate Functions
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.expressions.MutableAggregationBuffer
import org.apache.spark.sql.expressions.UserDefinedAggregateFunction
import org.apache.spark.sql.types._

object Myaverage extends UserDefinedAggregateFunction {
	// data types of input arguments of this aggregation function
	def inputSchema: StructType = 
}


// Type-Safe User-Defined Aggregate Functions

File: python_version/spark_SQL_DataFrames_Datasets.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Spark SQL, DataFrames, Datasets Guide

# SQL
	# SQL
	# Hive Table
	# other programming language
# Datasets
	# distributed collection of data
	# Scala, Java
# DataFrame
	# Datasets organized into named columns
	# Table
	# Data Frame in R
	# DataFrame in python




from pyspark.sql import SparkSession
from pyspark.sql import DataFrame
from pyspark.sql import Column
from pyspark.sql import Row
from pyspark.sql import GroupedData
from pyspark.sql import DataFrameNaFunctions
from pyspark.sql import DataFrameStatFunctions
from pyspark.sql import types
from pyspark.sql.types import *
from pyspark.sql import functions
from pyspark.sql import streaming
from pyspark.sql import Window


# *********************************************************
# SparkSession
# *********************************************************
spark = SparkSession \
	.builder \
	.master("local[4]")
	.appName("Python Spark SQL") \
	.config("spark.some.config.option", "some-value") \
	.getOrCreate()


# *********************************************************
# DataFrames
# *********************************************************
people = spark.read.json("D:/spark-2.3.2-bin-hadoop2.7/examples/src/main/resources/people.json")
department = spark.read.parquet("D:/spark-2.3.2-bin-hadoop2.7/examples/src/main/resources/users.parquet")
people.show()
department.show()

people.printSchema()
people.select("name").show()
people.select(people['name'], (people['age'] + 1).alias("age_new")).show()
people.filter(people['age'] > 21).show()
people.groupBy('age').count().show()


# *********************************************************
# SQL Queries
# *********************************************************
# Register the DataFrame as a SQL temporary view
people.createOrReplaceTempView("people")
sqlDF = spark.sql("SELECT * FROM people")
sqlDF.show()

# Global temporary view
people.createOrReplaceGlobalTempView("people_global")
sqlDF_2 = spark.sql("SELECT * FROM global_temp.people_global")
sqlDF_2.show()

# cross-session
spark.newSession().sql("SELECT * FROM global_temp.people_global").show()


# *********************************************************
# Datasets
# *********************************************************
""" {scala}
case class Person(name: String, age:Long)
val caseClassDS = Seq(Person("Andy", 32)).toDS()
caseClassDS.show()

val primitiveDS = Seq(1, 2, 3).toDS()
primitiveDS.map(_ + 1).collect()

val path = "D:/spark-2.3.2-bin-hadoop2.7/examples/src/main/resources/people.json"
val peopleDS = spark.read.json(path).as[Person]
peopleDS.show()
"""

# *********************************************************
# Spark SQL and RDD
# *********************************************************
# Inferring the schema using reflection
sc = spark.sparkContext

# Load a text file and convert each line to a Row
lines = sc.textFile("D:/spark-2.3.2-bin-hadoop2.7/examples/src/main/resources/people.txt")
parts = lines.map(lambda l: l.split(","))
people = parts.map(lambda p: Row(name = p[0], age = int(p[1])))

# Infer the schema, and register the DataFrame as a table
schemaPeople = spark.createDataFrame(people)
schemaPeople.createOrReplaceGlobalTempView("people")

# SQL can be run over DataFrame that have been registered as table
teenagers = spark.sql("SELECT name FROM people WHERE age >= 13 AND age <= 19")

# The results of SQL queries are DataFrame object, rdd returns the contents as an :class: pyspark.RDD of :class:Row
teenNames = teenagers.rdd.map(lambda p: "Name: " + p.name).collect()
for name in teenNames:
	print(name)

# ------------------------------------------

# Programmatically specifying the schema
sc = spark.sparkContext

# RDD
parts = lines.map(lambda l: l.split(","))
people = parts.map(lambda p: (p[0], p[1].strip()))

# Apply the schema to RDD
schemaString = "name age"
fields = [StructField(field_name, StringType(), True) for field_name in schemaString.split()]
schema = StructType(fields)
schemaPeople = spark.createDataFrame(people, schema)

# Create a temporary view using the DataFrame
schemaPeople.createOrReplaceTempView("people")
results = spark.sql("SELECT name FROM people")
results.show()

# *********************************************************
# Aggregations
# *********************************************************



# *********************************************************
# Data Sources
# *********************************************************





# *********************************************************
# Data Sources
# *********************************************************



# *********************************************************
# Data types
# *********************************************************
from pyspark.sql.types import NullType
from pyspark.sql.types import StringType
from pyspark.sql.types import BinaryType
from pyspark.sql.types import BooleanType
from pyspark.sql.types import DateType
from pyspark.sql.types import TimestampType
from pyspark.sql.types import DecimalType
from pyspark.sql.types import DoubleType
from pyspark.sql.types import FloatType
from pyspark.sql.types import ByteType
from pyspark.sql.types import IntegerType
from pyspark.sql.types import LongType
from pyspark.sql.types import ShortType
from pyspark.sql.types import ArrayType
from pyspark.sql.types import MapType
from pyspark.sql.types import StructField
from pyspark.sql.types import StructType

